File: sensors/IMU.py
from dataclasses import dataclass
import numpy as np
from sensors.SensorManager import SensorManager

@dataclass
class IMUMeasurement:
    time: float
    acceleration: np.ndarray  # 3D acceleration vector
    gyroscope: np.ndarray     # 3D angular velocity vector

class IMUManager(SensorManager):
    def __init__(self, *args):
        super().__init__(*args)
        self.measurements = []
    
    def register_measurement(self, time: float, acceleration: np.ndarray, gyroscope: np.ndarray):
        if not isinstance(acceleration, np.ndarray) or acceleration.shape != (3,):
            raise ValueError("Acceleration must be a 3-element vector.")
        
        if not isinstance(gyroscope, np.ndarray) or gyroscope.shape != (3,):
            raise ValueError("Gyroscope data must be a 3-element vector.")
        
        self.measurements.append(IMUMeasurement(time, acceleration, gyroscope))
    
    def get_latest_measurement(self):
        if not self.measurements:
            raise ValueError("No measurements recorded.")
        return self.measurements[-1]
    
    def get_measurements_in_time_range(self, start_time: float, end_time: float):
        return [m for m in self.measurements if start_time <= m.time <= end_time]

File: c++/src/sensors/imu.cc
#include "sensors/imu.hh"

#include <stdexcept>

namespace frc_sysid {

// Constructor: Forwards configuration to the SensorManager base class.
IMUManager::IMUManager(const RobotConfig &robot_config, const FieldConfig &field_config)
    : SensorManager(robot_config, field_config)
{}

// Register a new measurement by appending it to the internal vector.
void IMUManager::registerMeasurement(double time,
                                      const Eigen::Vector3d &acceleration,
                                      const Eigen::Vector3d &gyroscope)
{
  Measurement m;
  m.time = time;
  m.acceleration = acceleration;
  m.gyroscope = gyroscope;
  measurements.push_back(m);
}

// Returns the most recent measurement. Throws an exception if none exist.
const IMUManager::Measurement& IMUManager::getLatestMeasurement() const {
  if (measurements.empty())
    throw std::runtime_error("No measurements recorded.");
  return measurements.back();
}

// Returns all measurements whose time falls within the [start_time, end_time] range.
std::vector<IMUManager::Measurement> IMUManager::getMeasurementsInTimeRange(double start_time, double end_time) const {
  std::vector<Measurement> result;
  for (const auto &m : measurements) {
    if(m.time < start_time) continue;
    else if(m.time > end_time) break;

    result.push_back(m);
  }
  return result;
}

}